vignettes/package.Rmd

```diff
@@ -0,0 +1,349 @@
+#> ✓ Writing 'R/extendr-wrappers.R'
+#> ✓ Finished configuring extendr for package myextendr.
+#> • Please update the system requirement in DESCRIPTION file.
+#> • Please run `rextendr::document()` for changes to take effect.
+```
+
+Now we are just one step away (as the message says, we need to run `rextendr::document()`) from calling Rust fucntions from R.
+But, before moving forward, let's look at the files added.
+
+## Package structure
+
+The below files are the ones `rextendr::use_extendr()` added.
+
+```
+.
+├── R
+│   └── extendr-wrappers.R
+...
+└── src
+    ├── Makevars
+    ├── Makevars.win
+    ├── entrypoint.c
+    └── rust
+        ├── Cargo.toml
+        └── src
+            └── lib.rs
+```
+
+* **`R/extendr-wrappers.R`**: This file contains auto-generated R functions from Rust code. We don't modify this by hand.
+* **`src/Makevars`**, **`src/Makevars.win`**: This hooks `cargo build` at the installation of the R package. In most of the cases, we don't edit these.
+* **`src/entrypoint.c`**: This is needed to avoid the linker removing the static library. In 99.9% of the cases, we don't edit this (except for changing the crate name).
+* **`src/rust/`**: Rust code of a crate using extendr-api. This is where we mainly write code.
+
+So, in short, what we should really look at is only these two files:
+
+### `src/rust/Cargo.toml`
+
+
+``` toml
+[package]
+name = 'myextendr'
+version = '0.1.0'
+edition = '2018'
+
+[lib]
+crate-type = [ 'staticlib' ]
+
+[dependencies]
+extendr-api = '*'
+```
+
+The create name is the same name as the R package's name by default. You can change this,
+but it might be a bit tired to tweak other files accordingly, so I recommend leaving this.
+
+To try the dev version of the extendr, you can modify the last line to
+
+``` toml
+extendr-api = { git = 'https://github.com/extendr/extendr' }
+```
+
+### `src/rust/src/lib.rs`
+
+``` rs
+use extendr_api::prelude::*;
+
+/// Return string `"Hello world!"` to R.
+/// @export
+#[extendr]
+fn hello_world() -> &'static str {
+    "Hello world!"
+}
+
+// Macro to generate exports.
+// This ensures exported functions are registered with R.
+// See corresponding C code in `entrypoint.c`.
+extendr_module! {
+    mod myextendr;
+    fn hello_world;
+}
+```
+
+Let's explain this part by part.
+
+The first line `use extendr_api::prelude::*;` loads functions used frequently.
+
+Next, your eyes might notice the `/` are repeated 3 times, while the usual
+Rust comment requires only twice (i.e. `//`). These are treated as roxygen comments
+and copied to the auto-generated R code. This is analogous to Rcpp/cpp11's `//'`.
+
+``` rs
+/// Return string `"Hello world!"` to R.
+/// @export
+```
+
+The next line is the core of `extendr`'s mechanism. If the function is marked with this
+macro, the corresponding R function will be generated automatically (I'll explain
+the detail later). This is analogous to Rcpp's `[[Rcpp::export]]` and cpp11's `[[cpp11::register]]`.
```

Comment:
Remove "I'll epxlain the detail later". This vignette should only be about setting up packages. Other parts of extendr can be explained in other vignettes.

Reply:
So, should I remove "Rust code vs generated R code" section from this vignette?

Reply:
No, I wouldn't remove it. Just when I read "I'll explain later" it sounded to me as if the vignette was going to give a full-blown intro to how extendr works, and we don't need that here. If you want some sort of a pointer for future material there, maybe just write "... will be generated automatically (see below)". But no pointer at all works also.

Reply:
Ah, I see. Let's just remove confusing "I'll explain ...".